### PR TITLE
EOS-20373:Implementation for limit to create CORTX administrative accounts

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Backend

## Problem Statement
<pre>
Story Ref (if any):
</pre>
[EOS-20373](https://jts.seagate.com/browse/EOS-20373)
## Unit testing on RPM done
<pre>
Yes
</pre>
## Problem Description

- Limit to create CORTX administrative accounts for LR2
- Max users allowed = 100

## Solution
- Added new key in csm.conf to configure the value of max users allowed.
- Logic implemented to restrict account creation if maximum users limit reached.
## Unit Test Cases
- Account creation fails if maximum users limit reached.

1. On GUI
image

2. On Cortxcli
image

3. Postman
image

Signed-off-by: rohitkolapkar <rohit.j.kolapkar@seagate.com>

### CODE REVIEW

**Review by udayan-y21 — CHANGES_REQUESTED** [review]
Add screenshot for CLI execution and Postman API execution.

**Comment by rohitkolapkar** [comment]
Custom Ci Build 
http://eos-jenkins.mero.colo.seagate.com/job/GitHub-custom-ci-builds/job/centos-7.8/job/cortx-custom-ci/1582/

**Comment by rohitkolapkar** [comment]
> Add screenshot for CLI execution and Postman API execution.
> Add VM deployment link as well as customci build link

Hi, @udayan-y21 

- Screenshots for CLI execution and Postman execution added.
- Custom Ci build link added

Do I need to deploy VM manually and add deployment job link here ..?

**Comment by udayan-y21** [comment]
> > Add screenshot for CLI execution and Postman API execution.
> > Add VM deployment link as well as customci build link
> 
> Hi, @udayan-y21
> 
> * Screenshots for CLI execution and Postman execution added.
> * Custom Ci build link added
> 
> Do I need to deploy VM manually and add deployment job link here ..?

@rohitkolapkar As you know, automated miniprovisioning and vm deployment flags  is disabled deu to infra issue... so we need to test it mannually.. that will ensure that deployment isnt breaking deu to your code changes.

File: csm/core/services/users.py [hunk 1]
```diff
@@ -213,6 +213,12 @@ async def get_user_list(self, limit, offset, sort_by, sort_dir):
 
         return [self._user_to_dict(x) for x in user_list]
 
+    async def get_user_count(self):
+        """
+        returns the count of existing users
+        """
+        return await self.user_mgr.count()
```
Comment (avrnv) [inline csm/core/services/users.py:220]: Do you need to limit overall CSM users or only users with 'admin' role?

**Review by avrnv — CHANGES_REQUESTED** [review]

File: csm/core/services/users.py (hunk 1 above)
Comment (rohitkolapkar) [reply csm/core/services/users.py:220]: @avrnv 

As discussed with @pawankumarsrivastava, requirement is to limit all the CSM users ( role[admin,manager,monitor]).

**Review by pawankumarsrivastava — APPROVED** [review]

**Comment by rohitkolapkar** [comment]
> > > Add screenshot for CLI execution and Postman API execution.
> > > Add VM deployment link as well as customci build link
> > 
> > 
> > Hi, @udayan-y21
> > 
> > * Screenshots for CLI execution and Postman execution added.
> > * Custom Ci build link added
> > 
> > Do I need to deploy VM manually and add deployment job link here ..?
> 
> @rohitkolapkar As you know, automated miniprovisioning and vm deployment flags is disabled deu to infra issue... so we need to test it mannually.. that will ensure that deployment isnt breaking deu to your code changes.

@udayan-y21 
**VM Deployment passed** : http://eos-jenkins.colo.seagate.com/job/Cortx-Deployment/job/VM-Deployment-R2-1Node/997/console

File: csm/core/controllers/users.py [hunk 2]
```diff
@@ -154,6 +155,11 @@ async def post(self):
         except ValidationError as val_err:
             raise InvalidRequest(f"Invalid request body: {val_err}")
 
+        max_users_allowed = Conf.get(const.CSM_GLOBAL_INDEX, f"{const.CSM_MAX_USERS_ALLOWED}")
+        existing_users_count = await self._service.get_user_count()
+        if existing_users_count>=max_users_allowed:
+            raise CsmPermissionDenied("User creation failed. Maximum user limit reached.")
+
```
Comment (avrnv) [inline csm/core/controllers/users.py:162]: Minor: create a separate function for the check.

**Review by avrnv — CHANGES_REQUESTED** [review]

File: csm/core/controllers/users.py (hunk 2 above)
Comment (rohitkolapkar) [reply csm/core/controllers/users.py:165]: @avrnv  done

**Review by avrnv — APPROVED** [review]

**Pull request closed without merging**